src/hotspot/share/jvmci/jvmciCompiler.cpp

```diff
@@ -153,6 +153,36 @@ void JVMCICompiler::compile_method(ciEnv* env, ciMethod* target, int entry_bci,
   ShouldNotReachHere();
 }
 
+void JVMCICompiler::stopping_compiler_thread(CompilerThread* current) {
+  if (UseJVMCINativeLibrary) {
+    JVMCIRuntime* runtime = JVMCI::compiler_runtime(current, false);
+    if (runtime != nullptr) {
+      MutexUnlocker unlock(CompileThread_lock);
+      runtime->detach_thread(current, "stopping idle compiler thread");
+    }
+  }
+}
+
+void JVMCICompiler::on_empty_queue(CompileQueue* queue, CompilerThread* thread) {
+  if (UseJVMCINativeLibrary) {
+    int delay = JVMCICompilerIdleDelay;
+    JVMCIRuntime* runtime = JVMCI::compiler_runtime(thread, false);
+    // Don't detach JVMCI compiler threads from their JVMCI
+    // runtime during the VM startup grace period
+    if (runtime != nullptr && delay > 0 && tty->time_stamp().milliseconds() > DEFAULT_COMPILER_IDLE_DELAY) {
```

Comment:
Can it be more coordinated by some status instead of time? Time depends on machine you run and does not guarantee right case. I could be done in separate changes.

Reply:
This check exists only to avoid adding extra overhead of creating and destroying libgraal runtimes during early VM execution for very short lived apps (i.e. less than 1000 ms). After that, normal libgraal runtime management should not be a problem. In combination with [lazy allocation of compiler threads](https://bugs.openjdk.java.net/browse/JDK-8198756), it's very rare in practice for a compiler thread to want to shutdown anyway during the first 1000 ms of VM execution.

Reply:
Thank you for explanation.